### PR TITLE
Enable Tabix download + New datatype JuicerMediumTabix in 22.05

### PR DESCRIPTION
**Pull request by lldelisle**
I properly define the tabix datatype, enable download for this (I am not sure it is necessary).
I also define the JuicerMediumTabix which I use to generate matrices with cooler and that can be used by the juicer tools pre to generate '.hic'.

I need help to define good tests.
For the moment, I just added a small tabix file to test the sniff.

I tested locally that the vcf_bgzip can still be opened in IGV.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by lldelisle** [comment]
Mulled Unit Tests is failing with: https://github.com/galaxyproject/galaxy/actions/runs/3496584829/jobs/5854661557#step:8:641
```
    @external_dependency_management
    def test_get_alternative_versions():
        versions = get_alternative_versions("recipes/bamtools", "meta.yaml")
>       assert versions == ["recipes/bamtools/2.3.0/meta.yaml"]
E       AssertionError: assert [] == ['recipes/bam....0/meta.yaml']
E         Right contains one more item: 'recipes/bamtools/2.3.0/meta.yaml'
E         Full diff:
E         - ['recipes/bamtools/2.3.0/meta.yaml']
E         + []
```
Test Galaxy packages is failing in 3.10 with:
```
+ make mypy
mypy .
/tmp/gxpkgtestenvFHn7Gj/lib/python3.10/site-packages/numpy/__init__.pyi:638: error:
Positional-only parameters are only supported in Python 3.8 and greater 
[syntax]
        def seek(self, offset: int, whence: int, /) -> object: ...
                                                   ^
Found 1 error in 1 file (errors prevented further checking)
make: *** [Makefile:96: mypy] Error 2
ERROR: InvocationError for command /usr/bin/bash packages/test.sh (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   test_galaxy_packages: commands failed
Error: Process completed with exit code 1.
```
For 3.7 I get 2 errors:
```
=========================== short test summary info ============================
FAILED tests/webapps/test_middleware_empty_response.py::test_client_disconnect_with_middleware - AssertionError: assert None == 'No response returned.'
FAILED tests/webapps/test_middleware_empty_response.py::test_client_disconnect_raises_error_without_middleware - AssertionError: assert None == 'No response returned.'
================== 2 failed, 39 passed, 4 warnings in 25.67s ===================
```
Integration tests fail but this is my resposibility, I am pushing a fix.

**Comment by mvdbeek** [comment]
Yep, I think both of those tests are unrelated and fixed on dev

File: lib/galaxy/datatypes/tabular.py [hunk 1]
```diff
@@ -1779,3 +1779,90 @@ def set_meta(self, dataset, overwrite=True, skip=None, max_data_lines=7, **kwd):
...
+        # Check that the file is compressed with bgzip (not gzip), i.e. the
+        # compressed format is BGZF, as explained in
+        # http://samtools.github.io/hts-specs/SAMv1.pdf
+        with open(filename, "rb") as fh:
+            fh.seek(-28, 2)
+            last28 = fh.read()
+            return binascii.hexlify(last28) == b"1f8b08040000000000ff0600424302001b0003000000000000000000"
+
+    # Ideally the tabix_index would be regenerated when the metadataElements are updated
+    # As it is not the case for the moment, it is better not to set it wrongly
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/tabular.py]:
```suggestion
```

**Review by mvdbeek — APPROVED** [review]

**Comment by lldelisle** [comment]
@mvdbeek have you checked that UCSC visualization was still working for vcf_bgzip?
Do you think of other tests I could add?

**Comment by lldelisle** [comment]
@mvdbeek you were right. tabix is indeed used to store the tabix_index.
I thought I used interval_to_tabix converter while I used interval_to_bgzip converter when I was trying my workflows. This also means that my cooler tools are wrong as they use 'tabix' datatype wrongly...
Therefore the converter tests failed...
I go for 'RealTabix'? Can I keep the file_ext 'tabix'?

**Comment by lldelisle** [comment]
I made an attempt on https://github.com/lldelisle/galaxy/tree/real_tabix_22.05 let see how the tests go.

File: lib/galaxy/config/sample/datatypes_conf.xml.sample [hunk 2]
```diff
@@ -475,7 +475,8 @@
     <datatype extension="protobuf2" type="galaxy.datatypes.binary:Binary" subclass="true" description="Protocol Buffers (Protobuf) is data format for serializing structured data." display_in_upload="true"/>
     <datatype extension="protobuf3" type="galaxy.datatypes.binary:Binary" subclass="true" description="Protocol Buffers (Protobuf) is data format for serializing structured data." display_in_upload="true"/>
     <datatype extension="onnx" type="galaxy.datatypes.binary:Binary" subclass="true" description="ONNX (Open neural network exchange) is data format for storing and sharing machine learning and deep learning models." display_in_upload="true"/>
-    <datatype extension="tabix" type="galaxy.datatypes.binary:Binary" subclass="true"/>
+    <datatype extension="tabix" type="galaxy.datatypes.tabular:Tabix" mimetype="application/octet-stream" display_in_upload="true"/>
```
Comment (mvdbeek) [inline lib/galaxy/config/sample/datatypes_conf.xml.sample]:
```suggestion
    <datatype extension="tabix" type="galaxy.datatypes.binary:Binary" subclass="true"/>
```

**Comment by mvdbeek** [comment]
I was afraid that could be the case. 

I would suggest that we don't put real tabix in datatypes_conf.xml.sample and instead only provide the useful subclasses like `juicer.medium.tabix`. Eventually we should optimize the general purpose gz compression tool to generate valid tabix outputs for gff, gtf, bed etc, and we mark the compressed variants of these datatypes as having a tabix index.

File: lib/galaxy/config/sample/datatypes_conf.xml.sample [hunk 3]
```diff
@@ -1196,5 +1197,8 @@
     <sniffer type="galaxy.datatypes.binary:OxliSubset"/>
     <sniffer type="galaxy.datatypes.binary:OxliGraphLabels"/>
     <sniffer type="galaxy.datatypes.neo4j:Neo4jDBzip"/>
+    <sniffer type="galaxy.datatypes.tabular:VcfGz"/>
+    <sniffer type="galaxy.datatypes.binary:Bcf"/>
+    <sniffer type="galaxy.datatypes.tabular:Tabix"/>
```
Comment (mvdbeek) [inline lib/galaxy/config/sample/datatypes_conf.xml.sample]:
```suggestion
    <sniffer type="galaxy.datatypes.tabular:JuicerMediumTabix"/>
```

**Comment by lldelisle** [comment]
Let see the CI tests and I need to test it on my tools to see if it is good.

File: lib/galaxy/config/sample/datatypes_conf.xml.sample [hunk 4]
```diff
@@ -476,6 +476,10 @@
     <datatype extension="protobuf3" type="galaxy.datatypes.binary:Binary" subclass="true" description="Protocol Buffers (Protobuf) is data format for serializing structured data." display_in_upload="true"/>
     <datatype extension="onnx" type="galaxy.datatypes.binary:Binary" subclass="true" description="ONNX (Open neural network exchange) is data format for storing and sharing machine learning and deep learning models." display_in_upload="true"/>
     <datatype extension="tabix" type="galaxy.datatypes.binary:Binary" subclass="true"/>
+    <datatype extension="interval_tabix" type="galaxy.datatypes.interval:IntervalTabix" mimetype="application/octet-stream" />
+    <datatype extension="juicer_medium_tabix" type="galaxy.datatypes.interval:JuicerMediumTabix" mimetype="application/octet-stream" display_in_upload="true"/>
+    <datatype extension="bed_tabix" type="galaxy.datatypes.interval:BedTabix" mimetype="application/octet-stream" display_in_upload="true"/>
+    <datatype extension="gff_tabix" type="galaxy.datatypes.interval:GffTabix" mimetype="application/octet-stream" display_in_upload="true"/>
```
Comment (mvdbeek) [inline lib/galaxy/config/sample/datatypes_conf.xml.sample]: Is `.tabix` a common extension ? In that case I would suggest
```suggestion
    <datatype extension="interval.tabix" type="galaxy.datatypes.interval:IntervalTabix" mimetype="application/octet-stream" />
    <datatype extension="juicer_medium.tabix" type="galaxy.datatypes.interval:JuicerMediumTabix" mimetype="application/octet-stream" display_in_upload="true"/>
    <datatype extension="bed.tabix" type="galaxy.datatypes.interval:BedTabix" mimetype="application/octet-stream" display_in_upload="true"/>
    <datatype extension="gff.tabix" type="galaxy.datatypes.interval:GffTabix" mimetype="application/octet-stream" display_in_upload="true"/>
```
or `bed.tabix.gz` ? If we end with `.gz` tools that work with `.gz` will just work

Comment (lldelisle) [reply lib/galaxy/config/sample/datatypes_conf.xml.sample]: I think the common extension is `bgz` for bgzip (tabix only means that there is an associated tbi)... I noticed that all other extensions were with `_` that's why I put `_`...

**Comment by lldelisle** [comment]
I feel like with this version any gzip interval is recognized as interval_tabix while I remove the sniff to avoid it... Apparently it was not the good way. How can I prevent a random uploaded file to be recognized as interval_tabix?
Thanks

**Comment by lldelisle** [comment]
My tools seems to work with that version (https://github.com/lldelisle/tools-lldelisle/pull/5).
I just need to solve this sniffing.

File: lib/galaxy/config/sample/datatypes_conf.xml.sample (hunk 4 above)
Comment (mvdbeek) [reply lib/galaxy/config/sample/datatypes_conf.xml.sample]: For new extensions we usually do `{specific_extension}.{common_extension_that_osx/windows/linux_know_how_to_deal_with}`. All the auto-compressed datatypes for instance are `ext.gz`

File: lib/galaxy/datatypes/interval.py [hunk 5]
```diff
@@ -1729,6 +1731,114 @@ def sniff_prefix(self, file_prefix: FilePrefix):
...
+    MetadataElement(name="startCol", default=4, desc="Start column", param=metadata.ColumnParameter, readonly=True)
+    MetadataElement(name="endCol", default=4, desc="End column", param=metadata.ColumnParameter, readonly=True)
+
+
+class BedTabix(IntervalTabix):
+    """
+    Class describing a tabix file built from a bed file
+    """
+
+    file_ext = "bed_tabix"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/interval.py]: Considering that the samtools manpage uses `.gz` as the extension (http://www.htslib.org/doc/bgzip.html) I would say let's use `bed_tabix.gz` etc ?

File: lib/galaxy/datatypes/interval.py [hunk 6]
```diff
@@ -1729,6 +1731,114 @@ def sniff_prefix(self, file_prefix: FilePrefix):
...
+class IntervalTabix(Interval):
+    """
+    Class describing the bgzip format (http://samtools.github.io/hts-specs/SAMv1.pdf)
+    As tabix is just a bgzip sorted for chr,start,end with an index
+    """
+
+    file_ext = "interval_tabix"
+    edam_format = "format_3616"
+    compressed = True
+    compressed_format = "gzip"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/interval.py:1742]:
```suggestion
    compressed_format = "gzip"
    sniff = False
```

File: lib/galaxy/datatypes/interval.py [hunk 7]
```diff
@@ -1729,6 +1731,114 @@ def sniff_prefix(self, file_prefix: FilePrefix):
...
+    # We don't want to have a good sniff as the index would be created before the metadata on columns is set.
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        return False
+    #     # Check that the file is compressed with bgzip (not gzip), i.e. the
+    #     # compressed format is BGZF, as explained in
+    #     # http://samtools.github.io/hts-specs/SAMv1.pdf
+    #     with open(filename, "rb") as fh:
+    #         fh.seek(-28, 2)
+    #         last28 = fh.read()
+    #         return binascii.hexlify(last28) == b"1f8b08040000000000ff0600424302001b0003000000000000000000"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/interval.py]:
```suggestion
        if not self.sniff:
            return False
    # Check that the file is compressed with bgzip (not gzip), i.e. the
    # compressed format is BGZF, as explained in
    # http://samtools.github.io/hts-specs/SAMv1.pdf
    with open(filename, "rb") as fh:
         fh.seek(-28, 2)
         last28 = fh.read()
         return binascii.hexlify(last28) == b"1f8b08040000000000ff0600424302001b0003000000000000000000"
```

File: lib/galaxy/config/sample/datatypes_conf.xml.sample (hunk 4 above)
Comment (lldelisle) [reply lib/galaxy/config/sample/datatypes_conf.xml.sample]: OK then I will change

**Comment by lldelisle** [comment]
@mvdbeek if you find time to think about it...

**Review by mvdbeek — APPROVED** [review]

**Review by mvdbeek — APPROVED** [review]

**Comment by lldelisle** [comment]
Thank you

**Pull request merged**